ivadomed/loader/dataset_group.py

```diff
@@ -0,0 +1,159 @@
+        self.test_dataset = []
+
+        # These are lists of underlying pairing
+        self.train_filename_pairs: List[Tuple[list, list, str, dict]] = []
+        self.val_filename_pairs: List[Tuple[list, list, str, dict]] = []
+        self.test_filename_pairs: List[Tuple[list, list, str, dict]] = []
+
+        # DatasetGroup config
+        self.config = config
+
+        # This sets the expected number of files at the dataset group level. Useful for validating subsequent filesets and whether they are compliant or not.
+        self.n_expected_input: int = int(dict_file_group_spec.get(DataloaderKW.EXPECTED_INPUT, 0))
+        if self.n_expected_input <= 0:
+            raise ValueError("At DataSetGroup level, Number of Expected Input must be > 0")
+
+        self.n_expected_gt: int = int(dict_file_group_spec.get(DataloaderKW.EXPECTED_GT, 0))
+        if self.n_expected_gt <= 0:
+            raise ValueError("At DataSetGroup level, Number of Expected Ground Truth must be > 0")
+
+        # This is the key attribute that needs to be populated once data loading is complete.
+        if (ModelParamsKW.IS_2D not in config.model_params and not config.model_params.get(ModelParamsKW.IS_2D)):
+            self.validate_update_3d_train_val_test_and_filename_pairs(config, dict_file_group_spec)
+        else:
+            self.validate_update_2d_train_val_test_dataset_and_filename_pairs(config, dict_file_group_spec)
+
+        # Assign default labels if not provided.
+        self.name: str = dict_file_group_spec.get(DataloaderKW.DATASET_GROUP_LABEL, "DefaultDatasetGroupLabel")
+
+    def validate_update_3d_train_val_test_and_filename_pairs(self, config, dict_file_group_spec):
+        # Instantiate all the 3D Train/Val/Test Datasets
+        for a_train_dataset in dict_file_group_spec[DataloaderKW.TRAINING]:
+            if a_train_dataset[DataloaderKW.TYPE] == "FILES":
+                train_set = Files3DDataset(a_train_dataset, config)
+                self.train_dataset.append(train_set)
+                self.train_filename_pairs.extend(train_set.filename_pairs)
+        for a_val_dataset in dict_file_group_spec[DataloaderKW.VALIDATION]:
+            if a_val_dataset[DataloaderKW.TYPE] == "FILES":
+                val_set = Files3DDataset(a_val_dataset, config)
+                self.val_dataset.append(val_set)
+                self.val_filename_pairs.extend(val_set.filename_pairs)
+        for a_test_dataset in dict_file_group_spec[DataloaderKW.TEST]:
+            if a_test_dataset[DataloaderKW.TYPE] == "FILES":
+                test_set = Files3DDataset(a_test_dataset, config)
+                self.test_dataset.append(test_set)
+                self.test_filename_pairs.extend(test_set.filename_pairs)
+
+    def validate_update_2d_train_val_test_dataset_and_filename_pairs(self, config, dict_file_group_spec):
+        # Instantiate all the 2D Train/Val/Test Datasets
+        for a_train_dataset in dict_file_group_spec[DataloaderKW.TRAINING]:
```

Comment:
I think I'm missing something here. I thought that at this level, we were dealing with only one TRAIN/VALID/TEST list par group, so I'm not sure to understand why there are several TRAIN dataset in `dict_file_group_spec[DataloaderKW.TRAINING]`.

I would have expected to have a for loop going through TRAIN, VALID and TEST and not 3 different for loops.

Reply:
Great question. We technically support a LIST of training dataset. This way, train dataset could compose of many FIlesDataset/RegexDataset/BIdsDataset etc. All aggregated at the Datset Group level. Limit to a single TRAIN is a bit arbitrary and not flexible enough and the code is not that much complex (as you saw, three loops now instead of one loop). But underlying the dataset object work similar.

Reply:
Oh yeah, ok I get it, you could have a list of multiples FILES, or a FILES and a BIDS dataset, in the same Dataset Group.

I still wonder if that is necessary because we already have a list of DatasetGroup, so if we want to specify 2 different training lists, we could separate them in 2 DatasetGroup, no? With each training group being either FILES, BIDS or REGEX.

I'm saying this because right now we have 2 differents ways of dealing with this (IIUC). Let's say we have a FILES training set and a BIDS training set, we could:
1. Split the FILES and BIDS in different DatasetGroup
2. List them both in the training set of the same DatasetGroup.
From a user standpoint, if this flexibility is not absolutely required, it may be confusing to have 2 ways of doing this (and 2 ways of filling the config file).

Which brings me to yet another question:
It is possible to have only a training set in a DatasetGroup (no valid/test sets) or do we need to specify train/valid/test for all DatagsetGroup?

Reply:
> I still wonder if that is necessary because we already have a list of DatasetGroup, so if we want to specify 2 different training lists, we could separate them in 2 DatasetGroup, no? With each training group being either FILES, BIDS or REGEX.

Yes. Correct. Current design is flexible enough that however users want to do it, they can do it. The backend code does not really have strong preference either way. 

From what I have seen so far, I believe at DatasetGroup level, there is no requirement for validation/test set (given that we have mostly focused on the core training path). 

However, at the AllDatasetGroup level, I believe at least validation set is required for proper training.